src/category_theory/limits/shapes/products.lean

```diff
@@ -85,4 +85,11 @@ class has_coproducts :=
 
 attribute [instance] has_products.has_limits_of_shape has_coproducts.has_colimits_of_shape
 
+@[priority 100] -- see Note [lower instance priority]
+instance has_products_of_has_limits [has_limits C] : has_products C :=
+{ has_limits_of_shape := λ J, by apply_instance }
+@[priority 100] -- see Note [lower instance priority]
+instance has_coproducts_of_has_colimits [has_colimits C] : has_coproducts C :=
+{ has_colimits_of_shape := λ J, by apply_instance }
```

Comment:
Are these actually necessary? Don't they follow from the instances which get `has_products`  from `has_limits_of_shape (discrete J)` and `has_limits_of_shape` from `has_limits`?

Reply:
The problem is just that there's no such "instance which gets `has_products` from `has_limits_of_shape (discrete J)`.

I will check to see if I can make `has_products` just an abbreviation, rather than yet another ill-advised class.

Reply:
Yes, the abbreviation works fine, and saves half a dozen lines of boilerplate, so I've switched to that.